tests/utils/subnet.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (C) 2019-2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package utils
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"time"
+
+	"github.com/ava-labs/avalanchego/api/info"
+	"github.com/ava-labs/avalanchego/genesis"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/vms/secp256k1fx"
+	wallet "github.com/ava-labs/avalanchego/wallet/subnet/primary"
+	"github.com/ava-labs/subnet-evm/core"
+	"github.com/ava-labs/subnet-evm/plugin/evm"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/onsi/gomega"
+)
+
+func RunHardhatTests(test string, rpcURI string) {
+	log.Info("Sleeping to wait for test ping", "rpcURI", rpcURI)
+	client, err := NewEvmClient(rpcURI, 225, 2)
+	gomega.Expect(err).Should(gomega.BeNil())
+
+	bal, err := client.FetchBalance(context.Background(), common.HexToAddress(""))
+	gomega.Expect(err).Should(gomega.BeNil())
+	gomega.Expect(bal.Cmp(common.Big0)).Should(gomega.Equal(0))
+
+	err = os.Setenv("RPC_URI", rpcURI)
+	gomega.Expect(err).Should(gomega.BeNil())
+	cmd := exec.Command("npx", "hardhat", "test", fmt.Sprintf("./test/%s.ts", test), "--network", "local")
+	cmd.Dir = "./contract-examples"
+	log.Info("Running hardhat command", "cmd", cmd.String())
+
+	out, err := cmd.CombinedOutput()
+	fmt.Printf("\nCombined output:\n\n%s\n", string(out))
+	gomega.Expect(err).Should(gomega.BeNil())
```

Comment:
This is the ideal place to print an error if there is one. Makes debugging so so much easier.

Reply:
```
Combined output:



  ExampleTxAllowList
Contract deployed to: 0x52C84043CD9c865236f11d9Fc9F56aa003c1f922
    1) should add contract deployer as admin
    ✓ precompile should see admin address has admin role
    ✓ precompile should see test address has no role
    ✓ contract should report test address has no admin role
    ✓ contract should report admin address has admin role
    ✓ should not let test address submit txs (57ms)
    ✓ should not allow noRole to enable itself
    ✓ should not allow admin to enable noRole without enabling contract (58ms)
    ✓ should allow admin to add contract as admin (4166ms)
    ✓ should allow admin to add allowed address as allowed through contract (4107ms)
    ✓ should let allowed address deploy (4093ms)
    ✓ should not let allowed add another allowed
    ✓ should not let allowed to revoke admin
    ✓ should not let allowed to revoke itself
    ✓ should let admin to revoke allowed (4063ms)
    ✓ should not let admin to revoke itself


  15 passing (21s)
  1 failing

  1) ExampleTxAllowList
       should add contract deployer as admin:

      AssertionError: expected true to be false
      + expected - actual

      -true
      +false
      
      at Context.<anonymous> (test/tx_allow_list.ts:53:37)
      at step (test/tx_allow_list.ts:35:23)
      at Object.next (test/tx_allow_list.ts:16:53)
      at fulfilled (test/tx_allow_list.ts:7:58)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)




  [FAILED] Expected
      <*exec.ExitError | 0xc000436000>: {
          ProcessState: {
              pid: 11498,
              status: 256,
              rusage: {
                  Utime: {
                      Sec: 6,
                      Usec: 90941,
                      Pad_cgo_0: [0, 0, 0, 0],
                  },
                  Stime: {
                      Sec: 0,
                      Usec: 704411,
                      Pad_cgo_0: [0, 0, 0, 0],
                  },
                  Maxrss: 279552000,
                  Ixrss: 0,
                  Idrss: 0,
                  Isrss: 0,
                  Minflt: 96924,
                  Majflt: 1724,
                  Nswap: 0,
                  Inblock: 0,
                  Oublock: 0,
                  Msgsnd: 145,
                  Msgrcv: 145,
                  Nsignals: 0,
                  Nvcsw: 1850,
                  Nivcsw: 7595,
              },
          },
          Stderr: nil,
      }
  to be nil
  In [It] at: /Users/aaronbuchwald/go/src/github.com/ava-labs/subnet-evm/tests/utils/subnet.go:43 @ 01/26/23 15:20:49.916
  < Exit [It] tx allow list - /Users/aaronbuchwald/go/src/github.com/ava-labs/subnet-evm/tests/precompile/solidity/suites.go:36 @ 01/26/23 15:20:49.916 (27.514s)
• [FAILED] [27.514 seconds]
```

`CombinedOutput()` will print out stderr if it does run into an error, but gomega will still show the gnarly error from the hardhat command exiting with an error. This does give the line number that it failed on, which is helpful.

Here it is if it also prints the error:

```
  ExampleTxAllowList
Contract deployed to: 0x52C84043CD9c865236f11d9Fc9F56aa003c1f922
    1) should add contract deployer as admin
    ✓ precompile should see admin address has admin role
    ✓ precompile should see test address has no role
    ✓ contract should report test address has no admin role
    ✓ contract should report admin address has admin role
    ✓ should not let test address submit txs (38ms)
    ✓ should not allow noRole to enable itself
    ✓ should not allow admin to enable noRole without enabling contract
    ✓ should allow admin to add contract as admin (4077ms)
    ✓ should allow admin to add allowed address as allowed through contract (4057ms)
    ✓ should let allowed address deploy (4096ms)
    ✓ should not let allowed add another allowed
    ✓ should not let allowed to revoke admin
    ✓ should not let allowed to revoke itself
    ✓ should let admin to revoke allowed (4081ms)
    ✓ should not let admin to revoke itself


  15 passing (21s)
  1 failing

  1) ExampleTxAllowList
       should add contract deployer as admin:

      AssertionError: expected true to be false
      + expected - actual

      -true
      +false
      
      at Context.<anonymous> (test/tx_allow_list.ts:53:37)
      at step (test/tx_allow_list.ts:35:23)
      at Object.next (test/tx_allow_list.ts:16:53)
      at fulfilled (test/tx_allow_list.ts:7:58)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)





Err: exit status 1
  [FAILED] Expected
      <*exec.ExitError | 0xc000c80040>: {
          ProcessState: {
              pid: 12185,
              status: 256,
              rusage: {
                  Utime: {
                      Sec: 6,
                      Usec: 126482,
                      Pad_cgo_0: [0, 0, 0, 0],
                  },
                  Stime: {
                      Sec: 0,
                      Usec: 590972,
                      Pad_cgo_0: [0, 0, 0, 0],
                  },
                  Maxrss: 281849856,
                  Ixrss: 0,
                  Idrss: 0,
                  Isrss: 0,
                  Minflt: 97780,
                  Majflt: 1434,
                  Nswap: 0,
                  Inblock: 0,
                  Oublock: 0,
                  Msgsnd: 145,
                  Msgrcv: 145,
                  Nsignals: 0,
                  Nvcsw: 1392,
                  Nivcsw: 7641,
              },
          },
          Stderr: nil,
      }
  to be nil
  In [It] at: /Users/aaronbuchwald/go/src/github.com/ava-labs/subnet-evm/tests/utils/subnet.go:46 @ 01/26/23 15:24:48.054
  < Exit [It] tx allow list - /Users/aaronbuchwald/go/src/github.com/ava-labs/subnet-evm/tests/precompile/solidity/suites.go:36 @ 01/26/23 15:24:48.054 (27.512s)
```

Reply:
Adding it in makes it clearer that the hardhat process failed with exit status 1, but I think the part that was previously confusing was that the hardhat output did not show up.